src/util.cc

```diff
@@ -423,21 +427,82 @@ const char* SpellcheckString(const char* text, ...) {
 }
 
 #ifdef _WIN32
+std::wstring Utf8ToWide(const std::string& u8_string) {
+  if (u8_string.empty())
+    return std::wstring();
+
+  size_t size_needed =
+      MultiByteToWideChar(CP_UTF8, 0, u8_string.data(),
+                          static_cast<int>(u8_string.size()), NULL, 0);
+  if (size_needed == 0)
+    Fatal("Failed conversion to wide string");
+
+  vector<wchar_t> w_buffer(size_needed);
+  int w_size_needed = MultiByteToWideChar(CP_UTF8, 0, u8_string.data(),
+                                          static_cast<int>(u8_string.size()),
+                                          &w_buffer[0], w_buffer.size());
+  if (w_size_needed == 0)
+    Fatal("Failed conversion to wide string");
+
+  return std::wstring(&w_buffer[0], w_size_needed);
+}
+
+std::string WideToUtf8(const std::wstring& w_string) {
+  if (w_string.empty())
+    return std::string();
+
+  int size_needed = WideCharToMultiByte(CP_UTF8, 0, &w_string[0],
+                                        static_cast<int>(w_string.size()), NULL,
+                                        0, NULL, NULL);
+  if (size_needed == 0)
+    Fatal("Failed conversion to UTF-8 string");
+
+  std::string u8_string(size_needed, 0);
+  WideCharToMultiByte(CP_UTF8, 0, &w_string[0],
+                      static_cast<int>(w_string.size()), &u8_string[0],
+                      size_needed, NULL, NULL);
+  if (size_needed == 0)
+    Fatal("Failed conversion to UTF-8 string");
+
+  return u8_string;
+}
+
+char** convertCommandLine(int argc, wchar_t** wargv) {
+  std::wstring w_argString;
+  std::string argString;
+  char** argv = new char*[argc + 1];
+
+  for (int i = 0; i < argc; i++) {
+    // Convert the string
+    argString = WideToUtf8(wargv[i]);
```

Comment:
WideToUtf8 creates a full std::wstring.

You're allocating memory twice, and then throwing away the result.

Please have the Windows API place the converted string data *directly* into the destination where it will live for the rest of program runtime, without allocating or copying data unnecessarily.

Reply:
I've looked at this and is a bit unsure how to improve this. The string is needed when allocating space for the char* in the array. Could you elaborate more on plausible solutions?

Reply:
You don't need the actual std::wstring, you need the length for it.

You can query the length the same way you do inside the Utf8ToWide function, and then directly convert the data straight into the buffer that you allocate based on the length.

However, there are two better ways to handle this algorithm in general.

One is to replace the entire cli option handling of Ninja with a generic one that that lives in a template function that dynamically converts the arguments as needed, which doesn't seem like it would be included in this PR anyway,

and the other is this way: https://github.com/ninja-build/ninja/pull/1671/files/a6414b130eae643074ff5614c618222b96083c5e#r347538168

Quote:

> While I would strongly prefer to see this conversion to std::wstring also come with the reworked CLI parsing that I laid out in other comments:
> 
> If we're going to allocate a new char** array of arrays, and copy-convert the wchar_t** array of arrays into it, then do the following to help reduce the performance gap:
> 
> Calculate the total size of both the char** array, and also the individual char* arrays, including nul terminators
> 
> Allocate one buffer only, that encompasses the sizes of all items.
> 
> Populate the list of pointers to char, at the front of the newly allocated buffer
> Arrange the individual char arrays after the pointer list, and make them contiguous, with only the nul-terminator separating them in memory.
> 
> In this way, while we can't avoid copying each byte, we can avoid having to perform multiple allocations, representing a significant reduction on the number of blocking calls, which can have a big performance impact on heavily loaded systems


So in this way, you have
```cpp
char* totalBuffer = new char[argc*sizeof(char*) + totalLengthOfStringArgs];
char** retValue = static_cast<char**>(total_buffer);
for(each index)
{
    char* argAtIndex = total_buffer + argc*sizeof(char*) + index;
    // convert your Wide string to Utf8 directly into argAtIndex
    retValue[index] = argAtIndex;
}
return retValue;
```
Then, totalBuffer is garbage collected by the OS when the program terminates.

Reply:
The point here is that by doing it this way, you only pay for one allocation, which is much friendlier to the operating system when the computer is heavily loaded.

Considering Ninja's whole reason for existing was originally building the Chromium project, and building software tends to be pretty computer / memory heavy, that seems like a win.

Reply:
How about something like:
```
char** convertCommandLine(int argc, wchar_t** wargv) {
  size_t total_size = 0;

  for (int i = 0; i < argc; i++) {
    assert(!(wcslen(wargv[i]) > INT_MAX));

    total_size += WideCharToMultiByte(CP_UTF8, 0, wargv[i],
                                      static_cast<int>(wcslen(wargv[i])), NULL,
                                      0, NULL, NULL);

    total_size++;  // Add room for the null terminator.
  }

  size_t pointer_size = (argc + 1) * sizeof(char*);
  char* commandBuffer = new char[pointer_size + total_size];
  char** retvalue = reinterpret_cast<char**>(commandBuffer);
  int index = 0;

  for (int command_no = 0; command_no < argc; command_no++) {
    char* buf = pointer_size + commandBuffer + index;
    WideToUtf8(wargv[command_no], buf, total_size);
	index = index + strlen(buf) + 1; // Account for the null termination.
    retvalue[command_no] = buf;
  }

  // Just to follow the c++ standard, we need to add this!
  // Also, ninja will not work properly if we do not do so.
  retvalue[argc] = NULL;
  return retvalue;
}
```
This only does one allocation.

Reply:
A casual review of it doesn't show me any significant problems. So, yes, this is roughly what I had in mind.

Im curious why you call the WideToUtf8 function instead of the windows api function though. There doesn't seem to be any benefit to doing that, and I wonder about making sure your buffer lengths are properly accounted for.